### ISSUE TITLE
cmd/go: go get with ... fails in module mode

### ISSUE THREAD

**Issue by rsc** [golang/go#29363]
This used to work and should ideally keep working.

```
$ go get -u github.com/google/codesearch/cmd/...
go: finding github.com/google/codesearch/cmd/... latest
go: finding github.com/google/codesearch/cmd latest
go: finding github.com/google/codesearch latest
go: downloading github.com/google/codesearch v0.0.0-20150617151851-a45d81b686e8
go: extracting github.com/google/codesearch v0.0.0-20150617151851-a45d81b686e8
go get github.com/google/codesearch/cmd/...: no matching versions for query "latest"
$ go list github.com/google/codesearch/...
go: warning: "github.com/google/codesearch/..." matched no packages
```

**Comment by thepudds** [comment]
There might be some overlap with https://github.com/golang/go/issues/27215#issuecomment-427672781, which describes how  the documentation uses `go get golang.org/x/perf/cmd/...`, but that example fails (unless you have first done something like `go get golang.org/x/perf/cmd/benchstat`).

**Comment by hxzhao527** [comment]
the same:
```
go version
go version go1.11.4 linux/amd64
```
![image](https://user-images.githubusercontent.com/8402502/50731956-ed9b4800-11aa-11e9-9c54-4f91f47f1b18.png)

If no mod enabled,
![image](https://user-images.githubusercontent.com/8402502/50731961-fd1a9100-11aa-11e9-943b-96d0e9b53bcc.png)

I think this caused by: https://github.com/golang/go/blob/35f4ec152b44ae5fc83aaf68e2eb3aa1a778e5cd/src/cmd/go/internal/modfetch/coderepo.go#L97-L99
all versions are droped by the `prefix`, which is the `r.codeDir` . 

I add a `TrimSuffix` before https://github.com/golang/go/blob/35f4ec152b44ae5fc83aaf68e2eb3aa1a778e5cd/src/cmd/go/internal/modget/get.go#L347 to fix this. And it works.

```go
path = strings.TrimSuffix(path, "/...")
tasks = append(tasks, &task{arg: arg, path: path, vers: vers})
```

**Comment by hxzhao527** [comment]
I think this caued by this
https://github.com/golang/go/blob/master/src/cmd/go/internal/modload/import.go#L254-L270
when `go get -u github.com/google/codesearch/cmd/...`, dir is `$GOPATH/pkg/mod/github.com/google/codesearch@v1.1.0/cmd`. So `os.Open` will return `error`. And then *no go source* be throwed and `get` failed with not found.

I just motify this func to this:
```go
haveGoFiles = haveGoFilesCache.Do(dir, func() interface{} {
		if strings.HasSuffix(dir, "/..."){
			dir = strings.TrimSuffix(dir, "/...")
			tmp, err := filepath.Glob(dir+"/**/*.go")
			if err != nil{
				return false
			}
			return len(tmp) > 0
		}
		f, err := os.Open(dir)
                ...
```
And it works.

**Comment by mvdan** [comment]
@bcmills any chance this could be fixed in 1.12, or backported into a 1.12.x release at a later time?

It's great that with https://go-review.googlesource.com/c/go/+/148517, one can tell a program's users to install a specific version easily via `GO111MODULE=on go get foo.com/bar@v1.2.3`. However, it's a shame that `GO111MODULE=on go get foo.com/bar/cmd/...@v1.2.3` doesn't work.

I realise the plan is to fix it in 1.13, but that would mean easy install instructions would be delayed by six months, as one can't expect users to run on betas or master.

**Comment by bcmills** [comment]
@mvdan, it's not obvious to me whether the fix will be small or simple enough to backport.

There have been two independent attempts so far, and both have some issues in the details of the semantics and implementation. (The interaction between wildcard-expansion and module resolution is, unfortunately, quite subtle.)
